bittide/src/Bittide/DoubleBufferedRam.hs

```diff
@@ -247,7 +247,18 @@ wbStorage' initContent wbIn = delayControls wbIn wbOut
   delayControls m2s s2m0 = mux inCycle s2m1 (pure emptyWishboneS2M)
    where
     inCycle = (busCycle <$> m2s) .&&. (strobe <$> m2s)
-    delayedAck = register False (acknowledge <$> s2m0)
+    -- This is a workaround for an issue with the VexRiscv FFI.
+    -- The requests/responses have a one cycle delay, which means that it's
+    -- possible for the CPU to make a request, get no reply (because of the
+    -- delay) and then present that request *again*, even if the slave responds
+    -- directly.
+    -- The issue is that when the *next* request gets presented, the CPU would
+    -- see the acknowledge from a past request and get wrong data.
+    --
+    -- This workaround works by making sure there can be no two ACKs directly
+    -- after another. That forces the CPU to wait at least one cycle and
+    -- present/expect the right request/response.
+    delayedAck = register False (acknowledge <$> s2m0 .&&. (not <$> delayedAck))
     delayedErr = register False (err <$> s2m0)
     s2m1 = (\wb newAck newErr-> wb{acknowledge = newAck, err = newErr})
       <$> s2m0 <*> delayedAck <*> delayedErr
```

Comment:
```suggestion
    validRequest = (busCycle <$> m2s) .&&. (strobe <$> m2s)
    
    -- It takes a single cycle to lookup elements in a block ram. We can therfore
    -- only process a request every other clock cycle.
    ack = (not <$> ack) .&&. validRequest
    err1 = (err <$> s2m0) .&&. validRequest
    delayedAck = register False ack
    delayedErr1 = register False err1

    s2m1 = (\wb newAck newErr-> wb{acknowledge = newAck, err = newErr})
      <$> s2m0 <*> delayedAck <*> delayedErr1
```

Reply:
1. I've added ` .&&. validRequest` to `err`. I don't think you can flag an error if you're not in a valid cycle.
2. I don't really agree with how this code is structured overall. `go` is working in two clock cycles at the same time, it constructs an S2M which is later "fixed" using `delayControls`.
3. It's really weird that this issue wasn't caught by tests. Shouldn't reading from different addresses back-to-back raise issues?

I think we should restructure this code in a separate PR.

Reply:
Doesn't the `ack = (not <$> ack)` line introduce an evaluation loop?

Reply:
Whoops, yes

Reply:
1. there already was an `inCycle` variable that does exactly the same. I used that! I also changed the code to be `ack = (acknowledge <$> s2m0) .&&. (not <$> delayedAck) .&&. inCycle`, using only a cycling ACK doesn't catch when the storage didn't actually reply with an ACK one cycle ago.
2. :+1: 
3. Yes that should have been caught by tests. The test bench for the storage  that checks for *behaviour* doesn't use `wishbonePropWithModel` function from clash-protocols. I think we should maybe look into adding one of those and if that doesn't catch this error then modify clash-protocols so that this error gets caught!